### ISSUE TITLE
Get this error `Symbol not found: _GENERAL_NAME_free` at runtime in an electron application

### ISSUE THREAD

**Issue by zhaoyi0113** [grpc/grpc-node#624]
### What version of gRPC and what language are you using?
 grpc@1.13.1
 
### What operating system (Linux, Windows, …) and version?
 Darwin 192-168-1-6.tpgi.com.au 18.2.0 Darwin Kernel Version 18.2.0: Fri Oct  5 19:41:49 PDT 2018; root:xnu-4903.221.2~2/RELEASE_X86_64 x86_64
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 `yarn` and `npm`
 
### What did you do?
If possible, provide a recipe for reproducing the error. Try being specific and include code snippets if helpful.

 I am working on a vscode extension which is an electron application and when the application is launched, I see below errors:

```
dlopen(/Users/joeyzhao/dev/bigcrunch/extension/tbc/node_modules/grpc/src/node/extension_binary/node-v57-darwin-x64-unknown/grpc_node.node, 1): Symbol not found: _GENERAL_NAME_free
extensionHostProcess.js:621
  Referenced from: /Users/joeyzhao/dev/bigcrunch/extension/tbc/node_modules/grpc/src/node/extension_binary/node-v57-darwin-x64-unknown/grpc_node.node
  Expected in: flat namespace
 in /Users/joeyzhao/dev/bigcrunch/extension/tbc/node_modules/grpc/src/node/extension_binary/node-v57-darwin-x64-unknown/grpc_node.node.
```
### What did you see instead?
 
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md for how to diagnose problems better.

**Comment by nicolasnoble** [comment]
Please provide us with a reproduction case or repository.

**Comment by nicolasnoble** [comment]
How did you install node js on your macos machine?

**Comment by zhaoyi0113** [comment]
I am using `nvm` install `node v8.9.4`.

**Comment by nicolasnoble** [comment]
I'm really confused by this issue, since I don't really fully understand how everything works here, but I have the feeling this is a duplicate of mapbox/node-pre-gyp#432

You're installing the and *loading* the nodejs prebuild binary for grpc-node, whereas you're running in Electron. That's the overall issue here.

**Comment by asfaltboy** [comment]
So I found this same issue myself when trying to build grpc-node for a vscode extension.  It seems they use electron (3.1 now) but vscode specifies the target as node (currently node-v64-darwin-x64-unknown), because if I just build for electron I see this logged when I try to access grpc:

```
Activating extension 'undefined_publisher.my_extension' failed: Failed to load gRPC binary module because it was not installed for the current system
extensionHostProcess.js:735
Expected directory: node-v64-darwin-x64-unknown
Found: [electron-v3.0-darwin-x64-unknown, electron-v3.1-darwin-x64-unknown]
This problem can often be fixed by running "npm rebuild" on the current system
```

I think this is due to VSCode's extension host being spawned from an electron process as a node process (see [reference here][2]).

I found a workaround in another extension (blockchain-vscode-extension), to build grpc-node for electron, and rename to node (weird right?). See: [IBM-Blockchain/blockchain-vscode-extension#435][3] and [latest solution commit here][4]. I tried it quickly and it seems to load without issues, I still need to try it out but seems legit.

Once it runs on my dev machine I'll update ... but, how do we build this native code for all targets? Seems the IBM-Blockchain folks used something called [nativeDependencies][5], so maybe they handcrafted code to build it? Yes, check it out here: [DependencyManager.ts#L91][6].

[2]: https://github.com/Microsoft/vscode/issues/3011#issuecomment-184168434
[3]: https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues/435
[4]: https://github.com/caroline-church/blockchain-vscode-extension/commit/64db3762d6f718ecf3c16b709f8f7ed06b1c8feb
[5]: https://github.com/IBM-Blockchain/blockchain-vscode-extension/blob/a56e072111d008c134e4a1287aa5f82600c39042/client/package.json#L848
[6]: https://github.com/IBM-Blockchain/blockchain-vscode-extension/blob/c1836031b75512d79c07bc2b0b0ef3c77533d0f0/client/src/dependencies/DependencyManager.ts#L91

**Comment by nicolasnoble** [comment]
Would it be similar linked to mapbox/node-pre-gyp#278 ?

**Comment by nbransby** [comment]
Is there a workaround for using grpc in a vscode extension?

**Comment by asfaltboy** [comment]
Yes @nbransby , I managed to do it by building for electron and renaming the resulting binary lib to fit the node platform version expected by vscode's extension host. The downside that you would need to maintain this build/move script for each version change of vscode (or build multiple versions). See the original workaround I saw here: https://github.com/cazfletch/blockchain-vscode-extension/commit/64db3762d6f718ecf3c16b709f8f7ed06b1c8feb
